File: Top_Interview_150_List_With_solution(leetcode)/Array|String/L169-majority_element.py
# https://leetcode.com/problems/majority-element/
class Solution:
    def majorityElement(self, nums: List[int]) -> int:
        length, elem, count = len(nums), nums[0], 1
        n_by_two = length // 2
        for i in range(1, length):
            if (count == 0):
                count = 1
                elem = nums[i]
            elif (elem == nums[i]):
                count += 1
            else:
                count -= 1
        count = 0
        for i in range(length):
            if (nums[i] == elem):
                count += 1
            if (count > n_by_two):
                return elem
        return -1

File: Top_Interview_150_List_With_solution(leetcode)/Array|String/L169-MajorityElement.java
// https://leetcode.com/problems/majority-element/
class Solution {
    public int majorityElement(int[] nums) {
        int length = nums.length, nByTwo = length / 2, elem = nums[0],count = 1;
        for (int i = 1; i < length; i++) {
            if (count == 0) {
                count = 1;
                elem = nums[i];
            } else if (elem == nums[i])
                count++;
            else
                count--;
        }
        count = 0;
        for (int i = 0; i < length; i++) {
            if (nums[i] == elem)
                count++;
            if (count > nByTwo)
                return elem;
        }
        return -1;
    }
}
